portal-kernel/src/com/liferay/portal/kernel/servlet/PortletServlet.java

```diff
@@ -105,8 +109,20 @@ public void service(
 		portletSession.setHttpSession(session);
 
 		try {
-			PortletFilterUtil.doFilter(
-				portletRequest, portletResponse, lifecycle, filterChain);
+			if ((portletRequest instanceof HeaderRequest) &&
+				(portletResponse instanceof HeaderResponse)) {
+
+				if (filterChain instanceof HeaderFilterChain) {
```

Comment:
@ngriffin7a this just looks really weird to me, why this is not happening inside PortletFilterUtil? Just like other phases.

Reply:
@shuyangzhou See my reply regarding PortletFilterUtil.java -- we can't pass an instance of HeaderFilterChain to a method that takes a FilterChain.

Reply:
@shuyangzhou Hmm, there might be a problem with line#115 `filterChain instanceof HeaderFilterChain` since the assignment of `filterChain` above looks like this:
```
		FilterChain filterChain = (FilterChain)request.getAttribute(
			PORTLET_SERVLET_FILTER_CHAIN);
```
I will look at this the debugger and I will report back.

Reply:
@shuyangzhou OK, there's no problem with line#115 because FilterChainImpl implements both javax.portlet.HeaderFilterChain and javax.portlet.FilterChain